### PR TITLE
[parsers] json regex key bug fixes

### PR DESCRIPTION
**Pull request by jacknagz**
to: @ryandeivert 
cc: @airbnb/streamalert-maintainers @chunyong-lin 
size: medium

## Background

Follow up to #380 and #387.  

Deployment of JSON Regex Key matching against a larger set of data introduced more issues with parsing.  The loading of the regex matched string into `json.loads` threw errors since it was improperly caught.

## Changes

* Catch invalid matched JSON
* Add unit test coverage

## Testing

* Locally with unit tests

### CODE REVIEW

File: tests/unit/stream_alert_rule_processor/test_parsers.py [hunk 1]
```diff
@@ -90,6 +92,127 @@ class TestJSONParser(TestParser):
...
 
+    def test_schema_fail(self):
+        """JSON Parser - Schema Match Failure"""
+        # setup
+        options = {}
+        schema = self.config['logs']['unit_test_simple_log']['schema']
+        data = json.dumps({'hey': 1})
+
+        # get parsed data
+        parsed_data = self.parser_helper(data=data, schema=schema, options=options)
```
Comment (ryandeivert) [inline tests/unit/stream_alert_rule_processor/test_parsers.py]: There's a few areas where `options` is instantiated in the test as an empty dict and just passed along. The `parser_helper` function does a `get` with the keyword args for `options`, so you can safely omit `options=...` from the helper call altogether. Please update here and other places below where options is not used.

File: tests/unit/stream_alert_rule_processor/test_parsers.py [hunk 2]
```diff
@@ -90,6 +92,127 @@ class TestJSONParser(TestParser):
...
+            'host': 'my-host-1'
+        })
+
+        # get parsed data
+        parsed_data = self.parser_helper(data=data, schema=schema, options=options)
+        assert_false(parsed_data)
+
+    @patch('stream_alert.rule_processor.parsers.LOGGER')
+    def test_optional_keys_missing_in_schema(self, mock_logging):
+        """JSON Parser - Optional TLK Not in Schema"""
```
Comment (ryandeivert) [inline tests/unit/stream_alert_rule_processor/test_parsers.py]: TLK could be a bit confusing in test output, please use a human readable string here

File: tests/unit/stream_alert_rule_processor/test_parsers.py [hunk 3]
```diff
@@ -90,6 +92,127 @@ class TestJSONParser(TestParser):
...
+        })
+
+        # get parsed data
+        parsed_data = self.parser_helper(data=data, schema=schema, options=options)
+
+        assert_equal(len(parsed_data), 1)
+        assert_true(parsed_data[0]['streamalert:envelope_keys']['host'] == '')
+
+    @patch('stream_alert.rule_processor.parsers.LOGGER')
+    def test_regex_key_invalid_json(self, mock_logging):
```
Comment (ryandeivert) [inline tests/unit/stream_alert_rule_processor/test_parsers.py]: ++ for this test. Good thinking

File: tests/unit/stream_alert_rule_processor/test_parsers.py (hunk 1 above)
Comment (jacknagz) [reply tests/unit/stream_alert_rule_processor/test_parsers.py]: `parser_helper` doesn't do a `.get`, but I can make it do one

Comment (ryandeivert) [reply tests/unit/stream_alert_rule_processor/test_parsers.py]: sorry you're right it doesn't. you could also just pass `None` but either is fine

File: tests/unit/stream_alert_rule_processor/test_parsers.py [hunk 4]
```diff
@@ -90,6 +92,124 @@ class TestJSONParser(TestParser):
...
+        })
+
+        # get parsed data
+        parsed_data = self.parser_helper(data=data, schema=schema, options=options)
+
+        assert_true(mock_logging.debug.called)
+        assert_false(parsed_data)
+
+    def test_regex_key_non_json(self):
+        """JSON Parser - Regex Key with Invalid JSON Object"""
```
Comment (ryandeivert) [inline tests/unit/stream_alert_rule_processor/test_parsers.py]: I just noticed this test has the same description as the one above it. Can you update one of them please?

**Pull request merged**